dashboard/dashboard.go

```diff
@@ -0,0 +1,281 @@
+// Copyright 2017 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package dashboard
+
+//go:generate go-bindata -nometadata -o assets.go -prefix assets -pkg dashboard assets
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/ethereum/go-ethereum/p2p"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/rcrowley/go-metrics"
+	"golang.org/x/net/websocket"
+	"html/template"
+	"net"
+	"net/http"
+	"sync"
+	"sync/atomic"
+	"time"
+)
+
+const (
+	procSampleLimit = 200
+	memSampleLimit  = 200
+)
+
+var (
+	nextId uint32 = 0 // Next connection id
+)
+
+type dashboard struct {
+	config *Config
+
+	listener net.Listener
+	index    []byte    // Index page to serve up on the web
+	conns    []*client // Currently live websocket connections
+
+	mtrcs *mtrcs  `json:",omitempty"`
+	stats *status `json:",omitempty"`
+
+	lock sync.RWMutex // Lock protecting the dashboard's internals
+}
+
+type client struct {
+	conn   *websocket.Conn // Particular live websocket connection
+	logger log.Logger      // Logger for the particular live websocket connection
+}
+
+type mtrcs struct {
```

Comment:
`mtrcs`? No no no :P Don't do weird cryptic type names. If you want to avoid `metrics` due to a clash with the package name, that is fine, but never ever add such convoluted unix style names :P